cmd/hub/mcscontroller/controller.go

```diff
@@ -0,0 +1,169 @@
+)
+
+// ServiceExportReconciler reconciles a ServiceExport object
+type ServiceExportReconciler struct {
+	Client client.Client
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+}
+
+// +kubebuilder:rbac:groups=multicluster.x-k8s.io,resources=serviceexports,verbs=get;list;watch;update;patch
+// +kubebuilder:rbac:groups=multicluster.x-k8s.io,resources=serviceexports/finalizers,verbs=get;update;delete
+// +kubebuilder:rbac:groups=multicluster.x-k8s.io,resources=serviceexports/status,verbs=get;update
+// +kubebuilder:rbac:groups=multicluster.x-k8s.io,resources=serviceimports,verbs=create;get;list;update;patch;delete
+
+// Reconcile handles an update to a ServiceExport
+// By design, the event that triggered reconciliation is not passed to the reconciler
+// to ensure actions are based on state alone. This approach is referred to as level-based,
+// as opposed to edge-based.
+func (r *ServiceExportReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	log := r.Log.WithValues("serviceexport", req.NamespacedName)
+
+	// @todo extract and log the real (i.e., source) object name, namespace and cluster?
+	log.Info("Reconciling ServiceExport")
+
+	se := mcsv1a1.ServiceExport{}
+	if err := r.Client.Get(ctx, req.NamespacedName, &se); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Info("No ServiceExport found")
+		} else {
+			log.Error(err, "Error fetching ServiceExport")
+		}
+		return ctrl.Result{}, nil
+	}
+
+	result, err := r.reconcile(ctx, &se)
+	if err != nil {
+		return result, err
+	}
+
+	if se.GetDeletionTimestamp() != nil { // service export is marked to be deleted
+		return r.handleDelete(ctx, &se)
+	}
+
+	return result, err
+}
+
+// reconcile the cluster state.
+//
+// Reconciliation logic:
+// 1. list all relevant exports (i.e., having the same name and namespace labels)
+// 2. determine the primary export object
+// 3. iterate on all relevant export objects
+// 3a. if the object is being deleted (non empty deletion timestamp) - ignore
+// 3b. if object conflicts with primary - update export.Status.Condition.Conflict
+//		and delete corresponding ServiceImport.
+// 3c. otherwise, create/update the corrresponding ServiceImport and set the
+//		ServiceExport as owner.
+//
+// Notes:
+// - ensure objects are protected with a finalizer. Since we only receive the
+//	 affected object name into the reconciliation loop, we have no way to determine
+//	 and retrieve affected ServiceExports (i.e., client.List would not return the
+//   deleted object and we can't access its labels). The finalizer is cleared when we
+// 	 receive a notification for change in an object scheduled for deletion.
+// - ServiceImports are owned by the corresponding ServiceExport and are thus
+//	 automatically deleted by k8s when their ServiceExport is deleted.
+//
+func (r *ServiceExportReconciler) reconcile(ctx context.Context, se *mcsv1a1.ServiceExport) (ctrl.Result, error) {
```

Comment:
wouldn't the first thing to be done in `reconcile` is be to retrieve the ServiceExport? What do we get from passing in the NamespacedName explicitly instead of the SE we just retrieved?
😕

Reply:
another option is to move the finalizer check outside to the calling function

Reply:
Changed to pass in the original service and service name (cluster not currently consulted - could be added if needed).